File: book.py
"""Definition of the Book class."""
# Author(s): Pierre Abraham Mulamba
# Date of creation (modification): 20200224 (20200224)
# Usage: import book
#        create an object book = Book('GA403', 'Big C++', 2009, 8, 0, 0)

from dataclasses import dataclass


@dataclass(init=True, repr=True)
class Book(object):
    """
    Definition of a Book with the following attributes:
    quote
    title
    year
    minimal_age
    n_possess
    n_available.
    """
    nBooks = 0
    __slots__ = ['quote', 'title', 'year', 'minimal_age', 'n_possess', 'n_available']
    quote: str
    title: str
    year: int
    minimal_age: int
    n_possess: int
    n_available: int
    nBooks += 1

    # Accessors methods
    @property
    def get_quote(self) -> str:
        """
        Accesses and returns the quote attribute of the class Book
        :return: the attribute quote
        """
        return self.quote

    @property
    def get_title(self) -> str:
        """
        Accesses and returns the title attribute of the class Book
        :return: title
        """
        return self.title

    @property
    def get_year(self) -> int:
        """
        Accesses and returns the year attribute of the class Book
        :return: year
        """
        return self.year

    @property
    def get_minimal_age(self) -> int:
        """
        Accesses and returns the attribute minimal_age of the class Book
        :return: minimal_age
        """
        return self.minimal_age

    @property
    def get_n_possess(self) -> int:
        """
        Accesses and returns the attribute n_possess of the class Book
        :return: n_possess
        """
        return self.n_possess

    @property
    def get_n_available(self) -> int:
        """
        Accesses and returns the attribute n_available of the class Book
        :return: n_available
        """
        return self.n_available

    # Mutators methods
    def set_quote(self, quote):
        """
        Mutator method to set the attribute quote of the class Book
        :param quote: str
        :return: void
        """
        self.quote = quote

    def set_title(self, title):
        """
        Mutator method to set the attribute title of the class Book
        :param title: str
        :return: void
        """
        self.title = title

    def set_year(self, year):
        """
        Mutator method to set the attribute year of the class Book
        :param year: int
        :return: void
        """
        self.year = year

    def set_minimal_age(self, minimal_age):
        """
        Mutator method to set the attribute minimal_age of the class Book
        :param minimal_age: int
        :return: void
        """
        self.minimal_age = minimal_age

    def set_n_possess(self, n_possess):
        """
        Mutator method to set the attribute n_possess of the class Book
        :param n_possess:
        :return: void
        """
        self.n_possess = n_possess

    def set_n_available(self, n_available):
        """
        Mutator method to set the attribute n_available of the class Book
        :param n_available: int
        :return: void
        """
        self.n_available = n_available

    def info(self):
        """
        Method to access and return the attributes of the class Book
        :return: void:
        """
        return f"{self.get_quote}. {self.get_title}. {self.get_minimal_age} y.o."

File: Book.java
/*
 * To change this license header, choose License Headers in Project Properties.
 * To change this template file, choose Tools | Templates
 * and open the template in the editor.
 */
package com.smart.tech;

/**
 *
 * @author pamulamba
 */
public class Book extends Object{
    private String quote;
    private String title;
    private int year;
    private int minReaderAge;
    private int nPossess;
    private int nAvailables;
    private static int count;
    
    // Ctor
    public Book(){}
    
    public Book(String quote_, String title_, int year_, int n_possess){
        quote = quote_;
        title = title_;
        year = year_;
        nPossess = n_possess;
        nAvailables = nPossess;
        ++count;
    }
    
    /**
     *
     * @throws Throwable
     */
    @Override
    protected void finalize() throws Throwable{
        try {
            --count;
        }finally {
            super.finalize();
        }
    }
    
    // Accessor methods
    public String getQuote(){
        return quote;
    }
    
    public String getTitle(){
        return title;
    }
    
    public int getYear(){
        return year;
    }
    
    public int getMinReaderAge(){
        return minReaderAge;
    }
    
    public int getNPossess(){
        return nPossess;
    }
    
    public int getNAvailable(){
        return nAvailables;
    }
    
    public static int getCount(){
        return count;
    }
    
    // Copy ctor
    public Book(Book book){
        this.quote = book.getQuote();
        this.title = book.getTitle();
        this.year = book.getYear();
        this.minReaderAge = book.getMinReaderAge();
        this.nPossess = book.getNPossess();
        this.nAvailables = book.getNAvailable();
    }
    
    // Mutator methods
    public void setQuote(String quote_){
        quote = ((0 < quote_.length() && quote_.length() <= 20) ? quote_ : "");
    }
    
    public void setTitle(String title_){
        title = ((0 < title_.length() && title_.length() <= 20) ? title_ : "");
    }
    
    public void setYear(int year_){
        year = ((1600 < year_ && year_ <= 2020 ) ? year_: 9999);
    }
    
    public void setMinReaderAge(int min_reader_age){
        minReaderAge = ((3 < min_reader_age && min_reader_age <= 120) ? min_reader_age : 999);
    }
    
    public void setNPossess(int n_possess){
        nPossess = ((0 < n_possess && n_possess <= 100) ? n_possess : 0);
    }
    
    public void setNAvalaible(int n_availables){
        nAvailables = ((n_availables <= getNPossess()) ? n_availables : 0);
    }
    
    // Builder design pattern methods
    public Book quote(String quote){
        this.quote = quote;
        return this;
    }
    
    public Book title(String title){
        this.title = title;
        return this;
    }
    
    public Book year(int year){
        this.year = year;
        return this;
    }
    
    public Book minReaderAge(int min_reader_age){
        this.minReaderAge = min_reader_age;
        return this;
    }
    
    public Book nPossess(int n_possess){
        this.nPossess = n_possess;
        return this;
    }
    
    // Printing method
    public void print(){
        System.out.printf("Book: %s. %s. %d. %d y.o. and more\n",
                getQuote(), getTitle(), getYear(), getMinReaderAge());
    }

} // End class Book
